instrumentation/instagin/handler_test.go

```diff
@@ -0,0 +1,116 @@
+// (c) Copyright IBM Corp. 2021
+// (c) Copyright Instana Inc. 2016
+
+// +build go1.11
+
+package instagin_test
+
+import (
+	"net/http/httptest"
+	"reflect"
+	"testing"
+
+	"github.com/instana/testify/require"
+
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/instana/testify/assert"
+
+	"github.com/gin-gonic/gin"
+	instana "github.com/instana/go-sensor"
+	"github.com/instana/go-sensor/instrumentation/instagin"
+)
+
+func TestAddMiddleware(t *testing.T) {
+	const expectedHandlersAmount = 3
+
+	// create a gin engine with default handlers
+	engine := gin.Default()
+
+	handlerN1Pointer := reflect.ValueOf(engine.Handlers[0]).Pointer()
+	handlerN2Pointer := reflect.ValueOf(engine.Handlers[1]).Pointer()
+	assert.Len(t, engine.Handlers, expectedHandlersAmount-1)
+
+	// create a gin engine with default handlers and add middleware
+	engine = getInstrumentedEngine()
+
+	assert.Len(t, engine.Handlers, expectedHandlersAmount)
+
+	// check that middleware was added as a first handler
+	assert.NotEqual(t, handlerN1Pointer, reflect.ValueOf(engine.Handlers[0]).Pointer())
+	assert.Equal(t, handlerN1Pointer, reflect.ValueOf(engine.Handlers[1]).Pointer())
+	assert.Equal(t, handlerN2Pointer, reflect.ValueOf(engine.Handlers[2]).Pointer())
+}
+
+func TestResponseHeadersCollecting(t *testing.T) {
```

Comment:
Nitpicking: this test ensures propagation, not header collection 😉 The header collection would be populating the [`CollectableHTTPHeaders`](https://github.com/instana/go-sensor#capturing-custom-http-headers) option and checking that the entry span has corresponding values. You might want to do this together with checking the span content above.